### PR TITLE
feat: Implement category support

### PR DESCRIPTION
**Pull request by 0styx0**
This pr builds on and supersedes #323 to implement full category support using the `-c/—category` flag.
Solves #10

### CODE REVIEW

**Review by WhyNotHugo — CHANGES_REQUESTED** [review]
Thanks for addressing this, this is a pretty tricky one.

I think you're not deleting categories from the `categories` table when a todo is deleted -- and they might show up for later todos when ids get flushed.

**Comment by 0styx0** [comment]
> Thanks for addressing this, this is a pretty tricky one.
> 
> I think you're not deleting categories from the `categories` table when a todo is deleted -- and they might show up for later todos when ids get flushed.

the line `FOREIGN KEY(todos_id) REFERENCES todos(id) ON DELETE CASCADE` in https://github.com/0styx0/todoman/blob/b2063c84fbcd0fb84d3bb540e661e40f6b505d3b/todoman/model.py#L506 should take care of that. I added a test in 004d812c27d5a1ea2ec89f5c37bfdde50565f169 to make sure, let me know if I missed it though

File: todoman/cli.py [hunk 1]
```diff
@@ -404,7 +421,7 @@ def edit(ctx, id, todo_properties, interactive, raw):
 
     changes = False
     for key, value in todo_properties.items():
-        if value is not None:
+        if value:
```
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/cli.py:424]: reverted in 43d38cf8ca89881f41b81d4c3336b2fd4713ac48

File: todoman/formatters.py [hunk 2]
```diff
@@ -112,9 +118,10 @@ def compact_multiple(self, todos: Iterable[Todo], hide_list=False) -> str:
             # TODO: add spaces on the left based on max todos"
 
             # FIXME: double space when no priority
-            table.append(
-                f"[{completed}] {todo.id} {priority} {due} {recurring}{summary}"
-            )
+            # split into parts to satisfy linter line too long
+            part1 = f"[{completed}] {todo.id} {priority} {due} "
+            part2 = f"{recurring}{summary}{categories}"
+            table.append(part1 + part2)
```
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/formatters.py]: thanks! didn't know about that

File: todoman/model.py [hunk 3]
```diff
@@ -755,18 +782,23 @@ def todos(
             lists = [
                 list_.name if isinstance(list_, TodoList) else list_ for list_ in lists
             ]
-            q = ", ".join(["?"] * len(lists))
-            extra_where.append(f"AND files.list_name IN ({q})")
+            category_slots = ", ".join(["?"] * len(lists))
+            extra_where.append(f"AND files.list_name IN ({category_slots})")
```
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py:786]: that variable name seems to be the result of a search and replace that I did. reverted to previous content in a7341c331195f5d1ef4350d9f69894bac40574f3. 

with the revert, my code doesn't change that section, but in https://github.com/pimutils/todoman/blob/6460032da527cd6885621b882d9f37f653412feb/todoman/model.py#L754 (the current main's model.py) it uses list and not list_slots. if you'd still like me to change it then let me know

File: todoman/model.py [hunk 4]
```diff
@@ -810,15 +842,27 @@ def todos(
...
-        query = """
-              SELECT todos.*, files.list_name, files.path
-                FROM todos, files
-               WHERE todos.file_path = files.path {}
+        # TODO: check if works
+        if categories:
+            query = """
+            SELECT distinct todos.*, files.list_name, files.path
+            FROM todos, files, categories
+            WHERE categories.todos_id = todos.id and todos.file_path = files.path {}
```
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py:850]: the `SELECT distinct` should take care of that.
example:
```
❯ todo new test_todo -c one -c two
[ ] 54  2022-02-01 16:29 test_todo @misc [one, two]

todoman/tests on  implement_categories via 🐍 v3.10.2
❯ todo list -c one -c two
[ ] 54  2022-02-01 16:29 test_todo @misc [one, two]
```

[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py:850]: it seems as though if I create an index on categories.todos_id then joins would be performant

[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py:850]: I'm not super familiar with this, but it appears that the unique constraint I already have (`CONSTRAINT category_unique UNIQUE (todos_id,category)`) would create an index on todos_id, so I'll modify this query to perform a join.

[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py]: done in f81e26f0def69468a49b512fd7bdccec44685a00

File: todoman/model.py [hunk 5]
```diff
@@ -863,6 +907,18 @@ def _todo_from_db(self, row: dict) -> Todo:
...
+        query = """
+            SELECT distinct category
+            FROM categories
+            WHERE categories.todos_id = '{}'
+            """.format(
+            todo.id,
+        )
+        categories = self._conn.execute(query).fetchall()
+        todo.categories = [i["category"] for i in categories]
+
```
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/model.py]: fixed in 2825fbf371602cb509c0c9d69076765d0099b9cd

File: tests/test_model.py [hunk 6]
```diff
@@ -161,6 +172,43 @@ def test_retain_unknown_fields(tmpdir, create, default_database):
...
 
+def test_category_integrity(tmpdir, create, default_database):
+    create("test.ics", "UID:AVERYUNIQUEID\nSUMMARY:RAWR\n")
+    db = Database([tmpdir.join("default")], tmpdir.join("cache.sqlite"))
+
+    todo = db.todo(1, read_only=False)
+    todo.categories = ["hi", "hi"]
+
+    with pytest.raises(AlreadyExists):
+        default_database.save(todo)
```
Comment (WhyNotHugo) [inline tests/test_model.py:183]: This doesn't look correct: a todo with a category fails to save when it's edited?

File: tests/test_model.py [hunk 7]
```diff
@@ -161,6 +172,43 @@ def test_retain_unknown_fields(tmpdir, create, default_database):
...
+
+
+def test_category_deletes_on_todo_delete(tmpdir, create, default_database):
```
Comment (WhyNotHugo) [inline tests/test_model.py:186]: This test looks correct. :+1:

File: todoman/cli.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WhyNotHugo) [reply todoman/cli.py:424]: You changed the one in line 362, not this one.

**Review by WhyNotHugo — CHANGES_REQUESTED** [review]

File: tests/test_model.py (hunk 6 above)
Comment (0styx0) [reply tests/test_model.py:183]: It’s ensuring unique categories. The other option would be to silently ignore the duplicate, but I felt that was the wrong approach.

File: tests/test_cli.py [hunk 8]
```diff
@@ -27,7 +28,7 @@ def test_list(tmpdir, runner, create):
     assert not result.exception
     assert not result.output.strip()
 
-    create("test.ics", "SUMMARY:harhar\n")
+    create("test.ics", f"UID:{uuid4()}\nSUMMARY:harhar\n")
```
[the remark this answers is not in the document]
Comment (0styx0) [reply tests/test_cli.py:31]: The `group_concat` in https://github.com/pimutils/todoman/blob/118df5549043bc893893957b1445398ad0baccc5/todoman/model.py#L851 which is needed to `JOIN` the categories tag (as per https://github.com/pimutils/todoman/pull/475#discussion_r793606728) depends on grouping the todos by uid. This is only possible if they have a uid.

File: todoman/cli.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (0styx0) [reply todoman/cli.py]: How's this c6af1d7b32f668bddf40f2493fbd7b763fa2603f?

**Review by WhyNotHugo — APPROVED** [review]
LGTM. Sorry for the delay, lost sight of this.

**Pull request merged**